libmicrohttpd API change in 0.9.71

Hi! It's seems that a breaking change in libmicrohttpd's API is making collectd unable to compile with `-Werror`

---

*   Version of collectd: HEAD of main branch
*   Operating system / distribution: Arch Linux (or any other with libmicrohttpd 0.9.71)
*   Kernel version (if applicable): not applicable

## Expected behavior

Successful compilation of collectd

## Actual behavior

collectd cannot compile with `-Werror` and fails with

```
src/capabilities.c: In function ‘cap_start_daemon’:
src/capabilities.c:317:38: error: passing argument 5 of ‘MHD_start_daemon’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  317 |       flags, httpd_port, NULL, NULL, cap_http_handler, NULL,
      |                                      ^~~~~~~~~~~~~~~~
      |                                      |
      |                                      int (*)(void *, struct MHD_Connection *, const char *, const char *, const char *, const char *, size_t *, void **) {aka int (*)(void *, struct MHD_Connection *, const char *, const char *, const char *, const char *, long unsigned int *, void **)}

In file included from src/capabilities.c:35:
/usr/include/microhttpd.h:2428:45: note: expected ‘MHD_AccessHandlerCallback’ {aka ‘enum MHD_Result (*)(void *, struct MHD_Connection *, const char *, const char *, const char *, const char *, long unsigned int *, void **)’} but argument is of type ‘int (*)(void *, struct MHD_Connection *, const char *, const char *, const char *, const char *, size_t *, void **)’ {aka ‘int (*)(void *, struct MHD_Connection *, const char *, const char *, const char *, const char *, long unsigned int *, void **)’}
 2428 |                   MHD_AccessHandlerCallback dh, void *dh_cls,
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~^~

```
This is because an API breaking change in libmicrohttpd in version 0.9.71. The return value of callbacks is now `enum` in contrast to `int` for libmicrohttpd versions <0.9.71

## See

* Example of the [same case](
https://sourceware.org/git/?p=elfutils.git;a=commitdiff;h=acb453851c9e6c46531b70fda7396885c0e7e1db) taken from the elfutils project mail list where a working fix is shown.
* [libmicrohttpd changelog Wed 08 Apr 2020 10:53:01 PM CEST](https://fossies.org/linux/libmicrohttpd/ChangeLog)
> Introduce `enum MHD_Result` for #MHD_YES/#MHD_NO to avoid using 'int' so much.
> Note that this change WILL cause compiler warnings until (most) MHD callbacks
> in application code change their return type from 'int' to 'enum MHD_Result'.
> That said, avoiding possible confusions of different enums is going to make
> the code more robust in the future. For conditional compilation, test
> for "MHD_VERSION >= 0x00097002". -CG


## Steps to reproduce

*   Clone collectd in a system with libmicrohttpd 0.9.71
*   `build.sh && ./configure`
*   `make`

## Solution

Conditionalize the return type for callback functions inside `capabilities.c` and update `capabilities_test.c`

I'll be opening a PR soon fixing this. I would appreciate any feedback whatsoever. :)

--- a/src/capabilities.c
+++ b/src/capabilities.c
@@ -33,6 +33,11 @@
 #include "utils/dmi/dmi.h"
 
 #include <microhttpd.h>
+#if MHD_VERSION >= 0x00097002
+#define MHD_RESULT enum MHD_Result
+#else
+#define MHD_RESULT int
+#endif
 
 #include <jansson.h>
 #include <netdb.h>
@@ -187,7 +192,7 @@ static int cap_get_dmi_variables(json_t *parent, const dmi_type type,
 
 /* http_handler is the callback called by the microhttpd library. It essentially
  * handles all HTTP request aspects and creates an HTTP response. */
-static int cap_http_handler(void *cls, struct MHD_Connection *connection,
+static MHD_RESULT cap_http_handler(void *cls, struct MHD_Connection *connection,
                             const char *url, const char *method,
                             const char *version, const char *upload_data,
                             size_t *upload_data_size, void **connection_state) {

--- a/src/capabilities.c
+++ b/src/capabilities.c
@@ -193,9 +193,10 @@ static int cap_get_dmi_variables(json_t *parent, const dmi_type type,
 /* http_handler is the callback called by the microhttpd library. It essentially
  * handles all HTTP request aspects and creates an HTTP response. */
 static MHD_RESULT cap_http_handler(void *cls, struct MHD_Connection *connection,
-                            const char *url, const char *method,
-                            const char *version, const char *upload_data,
-                            size_t *upload_data_size, void **connection_state) {
+                                   const char *url, const char *method,
+                                   const char *version, const char *upload_data,
+                                   size_t *upload_data_size,
+                                   void **connection_state) {
   if (strcmp(method, MHD_HTTP_METHOD_GET) != 0) {
     return MHD_NO;
   }

--- a/src/capabilities_test.c
+++ b/src/capabilities_test.c
@@ -93,14 +93,14 @@ struct MHD_Response *MHD_create_response_from_data(size_t size, void *data,
   return mhd_res;
 }
 
-int MHD_add_response_header(struct MHD_Response *response, const char *header,
-                            const char *content) {
+MHD_RESULT MHD_add_response_header(struct MHD_Response *response,
+                                   const char *header, const char *content) {
   return 0;
 }
 
-int MHD_queue_response(struct MHD_Connection *connection,
-                       unsigned int status_code,
-                       struct MHD_Response *response) {
+MHD_RESULT MHD_queue_response(struct MHD_Connection *connection,
+                              unsigned int status_code,
+                              struct MHD_Response *response) {
   return MHD_HTTP_OK;
 }
 

